### ISSUE TITLE
--show-unbudgeted breaks budgeting subaccounts

### ISSUE THREAD

**Issue by wad209** [simonmichael/hledger#939]
Example ledger and budget:
```
~ monthly from 2019/01
    expenses:personal             $1,000.00
    liabilities

2019/01/01 Google home hub and home mini x2
    expenses:personal:electronics          $10.00
    liabilities                           $-10.00
```
Generating a budget report shows (as expected):
```
$ hledger balance --budget -f .hledger/test expenses
...
expenses:personal ||  $10.00 [   1% of $1,000.00] 
```
However, applying `--show-unbudgeted` shows:
```
 expenses:personal             ||       0 [   0% of $1,000.00] 
 expenses:personal:electronics ||      $10.00                  
```

the $10 electronics purchase should be included in the expenses:personal budget, but it reads as 0.

**Comment by wad209** [comment]
I suppose one could make an argument that expenses:personal:electronics is not budgeted per se, but I think the correct way would be to show it still (in case one wanted to budget it) but also to include the value in expenses:personal (as is done with normal --budget).

**Comment by simonmichael** [comment]
I think we can include the balance in expenses:personal, or display it as a separate item, but not both. I'm not quite sure what's most fitting at the moment.

**Comment by simonmichael** [comment]
Here's the doc for --show-unbudgeted:

> *By default, only accounts with budget goals during the report period are shown. --show-unbudgeted shows unbudgeted accounts as well. Top-level accounts with no budget goals anywhere below them are grouped under <unbudgeted>.*

I'm going to mark this as a bug, and would welcome a PR fixing it. Then a bit of real-world testing will help confirm it's the right fix.

**Comment by wad209** [comment]
If you wanted to point me in the general direction I could take a look.

**Comment by simonmichael** [comment]
Great! here you go: https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Reports/BudgetReport.hs#L63

**Comment by simonmichael** [comment]
I don't follow this entirely but I you may well be right. 

The `bal --budget` report is unlike all the others, eg showing mixed depths by default (by default showing a line item for each account with a budget regardless of depth and collapsing all the others). It appears to give a flat-ish report by default, unlike `bal` which is gives tree mode by default.

Here's the doc for budgetRollUp:
```
-- | Adjust a journal's account names for budget reporting, in two ways:
--
-- 1. accounts with no budget goal anywhere in their ancestry are moved 
--    under the "unbudgeted" top level account.
--
-- 2. subaccounts with no budget goal are merged with their closest parent account
--    with a budget goal, so that only budgeted accounts are shown. 
--    This can be disabled by --show-unbudgeted.
```

**Comment by simonmichael** [comment]
And from the --budget docs, http://hledger.org/hledger.html#budget-report:
```
By default, only accounts with budget goals during the report period are shown. --show-unbudgeted shows unbudgeted accounts as well. Top-level accounts with no budget goals anywhere below them are grouped under <unbudgeted>.
```
Perhaps it would be clearer if all unbudgeted accounts were under <unbudgeted>, not just top-level ones ? Or that might be weird, separating child accounts from their parents. We need more mockups showing proposed behaviour in such cases. [`tests/budget/budget.test`](https://github.com/simonmichael/hledger/blob/master/tests/budget/budget.test) is our current best understanding of what this report should do.

**Comment by holmescharles** [comment]
Hi, I think this issue is back with the newest version (1.24). 

Here's example output:

```
❯ hledger -f foo balance --budget -E -M --tree
Budget performance in 2019-01:

                               ||                       Jan
===============================++===========================
 expenses:personal             ||       0 [0% of $1,000.00]
 expenses:personal:electronics ||  $10.00
 liabilities                   || $-10.00 [1% of $-1000.00]
-------------------------------++---------------------------
                               ||       0 [              0]
```

where `foo` is:

```ledger

~ monthly from 2019/01
    expenses:personal             $1,000.00
    liabilities

2019/01/01 Google home hub and home mini x2
    expenses:personal:electronics          $10.00
    liabilities                           $-10.00
```

**Comment by Xitian9** [comment]
I think we need clarification about what correct behaviour is in this case. This case specifically is not in the tests, but there is a similar test and the specified behaviour is 

```

**Comment by holmescharles** [comment]
The issue seems to be related to the number of levels. It looks like Xitian9's post shows a max depth of 2, but my last post has a max depth of 3.

Actually, all you have to do is add a single cent to the account above and it works as in Xitian9's post.

```
❯ hledger -f foo balance --budget -E -M --tree
Budget performance in 2019-01:

                   ||                       Jan
===================++===========================
 expenses:personal ||  $10.01 [1% of $1,000.00]
   electronics     ||  $10.00
 liabilities       || $-10.01 [1% of $-1000.00]
-------------------++---------------------------
                   ||       0 [              0]
```

where foo is:

```
# vim: ft=ledger

~ monthly from 2019/01
    expenses:personal             $1,000.00
    liabilities


2019/01/01 Google home hub and home mini x2
    expenses:personal          $0.01
    liabilities

2019/01/01 Google home hub and home mini x2
    expenses:personal:electronics          $10.00
    liabilities                           $-10.00
```

TLDR, we can argue about whether or not account values should propagate up or not, but they should at least do it consistently. It should *not* depend on whether or not the parent account was referenced directly in another transaction.

**Comment by Xitian9** [comment]
Ah, I understand now. In fact, the issues here seems to be slightly broader: you'll notice that it's not even grouping the parent accounts in terms of layout as you would expect in tree mode.

**Comment by holmescharles** [comment]
Yes. Also the issues do not seem to be specific to tree mode. If you remove "--tree" from both of my previoud posts they yield:

Without the 1 cent expense:
```

                               ||                       Jan
===============================++===========================
 expenses:personal             ||       0 [0% of $1,000.00]
 expenses:personal:electronics ||  $10.00
 liabilities                   || $-10.00 [1% of $-1000.00]
-------------------------------++---------------------------
                               ||       0 [              0]
```

With it:

```
                               ||                       Jan
===============================++===========================
 expenses:personal             ||  $10.01 [1% of $1,000.00]
 expenses:personal:electronics ||  $10.00
 liabilities                   || $-10.01 [1% of $-1000.00]
-------------------------------++---------------------------
                               ||       0 [              0]
```

**Comment by Xitian9** [comment]
It seems to be a problem with boring parent elision not being able to talk between the actual and budget parts of the report. Notice that the problem disappears if you use `--no-elide`.

**Comment by holmescharles** [comment]
Great! So I have a work around for now. Though I hope this still gets flagged as a bug and is fixed. Thanks!

**Comment by Xitian9** [comment]
Definitely a bug. I'll try to get a fix out soon.

**Comment by simonmichael** [comment]
This continued and was resolved as #1800. Thanks!
